### PR TITLE
Add Product API to User timezone module

### PR DESCRIPTION
**Pull request by andretothie**
Adding the recommendation block, the recommendation controller, and the recommendation routing

### CODE REVIEW

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The query factory.
+   *
+   * @var \Drupal\Core\Config\Entity\QueryFactory
+   */
+  protected $queryFactory;
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]: You don't need this

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function index() {
+    return new JsonResponse([
+      'data' => $this->getData(),
+      'method' => 'GET',
+      'status' => 200,
+    ]
+    );
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
    return new JsonResponse($this->getData());
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function getData(): array {
+    $result = [];
+    $query = $this->entityTypeManager->getStorage('node')->getQuery();
+    $nids = $query
+      ->condition('type', 'article')
+      ->condition('status', TRUE)
+      ->condition('field_tags.entity.name', 'session time')
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]: Why is this condition needed? I think you want to do something else here.

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      ->execute();
+
+    if ($nids) {
+      foreach ($nids as $node_id) {
+        $node = $this->entityTypeManager->getStorage('node_id');
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]: Move the loading outside of the foreach. You want to load all nodes at the same time, like this:

```php
$nodes = $this->entityTypeManager->getStorage('node')->loadMultiple($nids);
```

Then you foreach over the `$nodes` array to construct the `$result` array.

**Review by tamasd — CHANGES_REQUESTED** [review]

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    $nids = $query
+      ->condition('type', 'article')
+      ->condition('status', '1')
+      ->condition('field_session_time.entity.name', $this->getTaxonomyName())
+      ->execute();
+
+    if ($nids) {
+      $node = $this->entityTypeManager->getStorage('node')->loadMultiple($nids);
+      array_rand($node);
+      $firstItem = reset($node);
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
      $firstItem = $node[array_rand($node, 1)];
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ * Class UserTimezoneRecommendationController.
+ */
+class UserTimezoneRecommendationController extends ControllerBase {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]: You don't need the whole entity type manager, you can inject just the node storage.

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+
+/**
+ * Class UserTimezoneRecommendationController.
+ */
+class UserTimezoneRecommendationController extends ControllerBase {
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]: The class can be final

File: web/modules/custom/user_timezone/src/Plugin/Block/UserTimezoneRecommendationBlock.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Provides UserTimezone Recommendation block.
+ *
+ * @Block(
+ *   id = "user_timezone_recommendation_block",
+ *   admin_label = @Translation("User timezone recommendation block"),
+ * )
+ */
+class UserTimezoneRecommendationBlock extends BlockBase implements ContainerFactoryPluginInterface {
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Plugin/Block/UserTimezoneRecommendationBlock.php]: This class can be final

File: web/modules/custom/user_timezone/src/Plugin/Block/UserTimezoneRecommendationBlock.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      $plugin_id,
+      $plugin_definition,
+      $container->get('user_timezone.recommendation')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Plugin/Block/UserTimezoneRecommendationBlock.php]: Missing return type `array`

File: web/modules/custom/user_timezone/src/UserTimezoneRecommendation.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+namespace Drupal\user_timezone;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\StringTranslation\TranslationInterface;
+
+/**
+ * Prepares the recommendation.
+ */
+class UserTimezoneRecommendation {
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/UserTimezoneRecommendation.php]: This class can be final

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      ->execute();
+
+    if ($nids) {
+      $node = $this->entityTypeManager->getStorage('node')->loadMultiple($nids);
+      array_rand($node);
+      $firstItem = reset($node);
+      return [
+        'Product name' => $firstItem->title->value,
+        'url' => $firstItem->toUrl()->toString(),
+      ];
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
      return new JsonResponse([
        'Product name' => $firstItem->label(),
        'url' => $firstItem->toUrl()->toString(),
      ]);
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function index(): array {
+    $json_array = [
+      'data' => [],
+    ];
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+      $node = $this->entityTypeManager->getStorage('node')->loadMultiple($nids);
+      array_rand($node);
+      $firstItem = reset($node);
+      return [
+        'Product name' => $firstItem->title->value,
+        'url' => $firstItem->toUrl()->toString(),
+      ];
+    }
+
+    return new JsonResponse($json_array);
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
    return new JsonResponse(new stdClass());
```

**Review by tamasd — CHANGES_REQUESTED** [review]

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+  /**
+   * Constructs the UserTimezoneRecommendationController object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\user_timezone\UserTimezoneSalutation $salutation
+   *   The salutation.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, UserTimezoneSalutation $salutation) {
+    $this->entityTypeManager = $entityTypeManager;
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
  public function __construct(EntityStorageInterface $nodeStorage, EntityStorageInterface $termStorage, UserTimezoneSalutation $salutation) {
    $this->nodeStorage = $nodeStorage;
    $this->termStorage = $termStorage;
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+ */
+final class UserTimezoneRecommendationController extends ControllerBase {
+
+  /**
+   * The recommendation.
+   *
+   * @var \Drupal\user_timezone\UserTimezoneSalutation
+   */
+  protected $salutation;
+
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php:24]:
```suggestion
private $nodeStorage;
private $termStorage;
```

and add proper comments

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      $container->get('entity_type.manager'),
+      $container->get('user_timezone.salutation')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function content(): JsonResponse {
+    $query = $this->entityTypeManager->getStorage('node')->getQuery();
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
    $query = $this->nodeStorage->getQuery();
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    $nids = $query
+      ->condition('type', 'product')
+      ->condition('status', '1')
+      ->condition('field_minimum_game_session_time.target_id', $this->getTaxonomyId())
+      ->execute();
+
+    if ($nids) {
+      $node = $this->entityTypeManager->getStorage('node')->loadMultiple($nids);
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
      $node = $this->nodeStorage->loadMultiple($nids);
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('user_timezone.salutation')
+    );
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
    $entityTypeManager = $container->get('entity_type.manager');
    return new static(
      $entityTypeManager->getStorage('node'),
      $entityTypeManager->getStorage('taxonomy_term'),
      $container->get('user_timezone.salutation')
    );
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+    elseif ($productRecommendation == (string) $this->t('Good evening')) {
+      $taxonomy = 'more than 1 h';
+    }
+
+    else {
+      $taxonomy = 'more than 1 h';
+    }
+
+    $terms = $this->entityTypeManager->getStorage('taxonomy_term')->loadByProperties([
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
    $terms = $this->termStorage->loadByProperties([
```

File: web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+      'name' => $taxonomy,
+      'vid' => 'session_time',
+    ]);
+
+    return $terms ? (int) reset($terms)->id() : NULL;
+
```
Comment (tamasd) [inline web/modules/custom/user_timezone/src/Controller/UserTimezoneRecommendationController.php]:
```suggestion
```

**Review by tamasd — CHANGES_REQUESTED** [review]

**Review by tamasd — APPROVED** [review]

**Review by kittiradovics — APPROVED** [review]

**Pull request merged**